src/dmd/ctfe/bc.d

```diff
@@ -0,0 +1,2369 @@
+            emitArithInstruction(LongInst.Xor32, result, rhs);
+        else
+            emitArithInstruction(LongInst.Xor, result, rhs);
+    }
+
+    void Lsh3(BCValue result, BCValue lhs, BCValue rhs)
+    {
+        assert(result.vType != BCValueType.Immediate, "Cannot lsh to Immediate");
+
+        result = (result ? result : lhs);
+        if (lhs != result)
+        {
+            Set(result, lhs);
+        }
+        emitArithInstruction(LongInst.Lsh, result, rhs);
+    }
+
+    void Rsh3(BCValue result, BCValue lhs, BCValue rhs)
+    {
+        assert(result.vType != BCValueType.Immediate, "Cannot rsh to Immediate");
+
+        result = (result ? result : lhs);
+        if (lhs != result)
+        {
+            Set(result, lhs);
+        }
+        emitArithInstruction(LongInst.Rsh, result, rhs);
+    }
+
+    void Mod3(BCValue result, BCValue lhs, BCValue rhs)
+    {
+        assert(result.vType != BCValueType.Immediate, "Cannot mod to Immediate");
+
+        result = (result ? result : lhs);
+        if (lhs != result)
+        {
+            Set(result, lhs);
+        }
+        emitArithInstruction(LongInst.Mod, result, rhs, &result.type.type);
+    }
+
+    void Umod3(BCValue result, BCValue lhs, BCValue rhs)
+    {
+        assert(result.vType != BCValueType.Immediate, "Cannot mod to Immediate");
+
+        result = (result ? result : lhs);
+        if (lhs != result)
+        {
+            Set(result, lhs);
+        }
+        emitArithInstruction(LongInst.Umod, result, rhs, &result.type.type);
+    }
+
+    void Call(BCValue result, BCValue fn, BCValue[] args, Loc l = Loc.init)
+    {
+        auto call_id = pushOntoStack(imm32(callCount + 1)).stackAddr;
+        calls[callCount++] = RetainedCall(fn, args, functionId, ip, sp, l);
+        emitLongInst(LongInst.Call, result.stackAddr, call_id);
+    }
+
+    void Load8(BCValue _to, BCValue from)
+    {
+        if (!isStackValueOrParameter(from))
+        {
+            from = pushOntoStack(from);
+        }
+        if (!isStackValueOrParameter(_to))
+        {
+            _to = pushOntoStack(_to);
+        }
+        assert(isStackValueOrParameter(_to), "to has the vType " ~ enumToString(_to.vType));
+        assert(isStackValueOrParameter(from), "from has the vType " ~ enumToString(from.vType));
+
+        emitLongInst(LongInst.HeapLoad8, _to.stackAddr, from.stackAddr);
+    }
+
+    void Store8(BCValue _to, BCValue value)
+    {
+        if (!isStackValueOrParameter(value))
+        {
+            value = pushOntoStack(value);
+        }
+
+        if (!isStackValueOrParameter(_to))
+        {
+            _to = pushOntoStack(_to);
+        }
+
+        assert(isStackValueOrParameter(_to), "to has the vType " ~ enumToString(_to.vType));
+        assert(isStackValueOrParameter(value), "value has the vType " ~ enumToString(value.vType));
+
+        emitLongInst(LongInst.HeapStore8, _to.stackAddr, value.stackAddr);
+    }
+
+    void Load16(BCValue _to, BCValue from)
```

Comment:
these `LoadNN` and `StoreNN` look very similar, there should probably be a common definition.